### PR TITLE
MAINT: Ensure that parsing errors are passed on even in tests.

### PR DESCRIPTION
**Pull request by mhvk**
To avoid it wasting a few hours of someone else's time if they edit the gufunc signature parser....

cc @mattip, since he may just have suffered similar problems...

### CODE REVIEW

File: numpy/core/src/umath/_umath_tests.c.src [hunk 1]
```diff
@@ -280,13 +280,21 @@ addUfuncs(PyObject *dictionary) {
                     "inner on the last dimension and broadcast on the rest \n"
                     "     \"(i),(i)->()\" \n",
                     0, inner1d_signature);
+    /* yes, this should not happen, but I just spent an hour looking for
```
Comment (charris) [inline numpy/core/src/umath/_umath_tests.c.src]: Multiline comments like so:
```
/*
 * blah
 * blah
 */
```

Comment (mhvk) [reply numpy/core/src/umath/_umath_tests.c.src]: done

**Review by mattip — APPROVED** [review]
LGTM

**Comment by eric-wieser** [comment]
Can you change `addUfuncs` to return an error code too?

File: numpy/core/src/umath/_umath_tests.c.src [hunk 2]
```diff
@@ -280,13 +280,23 @@ addUfuncs(PyObject *dictionary) {
                     "inner on the last dimension and broadcast on the rest \n"
                     "     \"(i),(i)->()\" \n",
                     0, inner1d_signature);
+    /*
+     * yes, this should not happen, but I (MHvK) just spent an hour looking at
```
Comment (eric-wieser) [inline numpy/core/src/umath/_umath_tests.c.src]: Not sure this comment is necessary - any `PyObject *` allocation can fail due to memory limits, so we should always check results like this.

File: numpy/core/src/umath/_umath_tests.c.src [hunk 3]
```diff
@@ -372,10 +391,12 @@ static struct PyModuleDef moduledef = {
 #endif
 
 #if defined(NPY_PY3K)
-#define RETVAL m
+#define RETVAL_SUCCESS m
+#define RETVAL_FAILURE NULL
```
Comment (eric-wieser) [inline numpy/core/src/umath/_umath_tests.c.src]: In `multiarraymodule.c` we use
```
#if defined(NPY_PY3K)
#define RETVAL(x) x
PyMODINIT_FUNC PyInit_multiarray(void) {
#else
#define RETVAL(x)
PyMODINIT_FUNC initmultiarray(void) {
#endif
```

would be nice to be consistent here

**Review by eric-wieser — CHANGES_REQUESTED** [review]
A clear improvement, but I think we could be a little more consistent with how we do things elsewhere here

File: numpy/core/src/umath/_umath_tests.c.src (hunk 2 above)
Comment (mhvk) [reply numpy/core/src/umath/_umath_tests.c.src]: Well, I'll maybe leave this as a breadcrump for the future - or to remind myself of hours waster...

File: numpy/core/src/umath/_umath_tests.c.src (hunk 3 above)
Comment (mhvk) [reply numpy/core/src/umath/_umath_tests.c.src]: Wasn't aware of that - like it much better, so will change.

File: numpy/core/src/umath/_umath_tests.c.src [hunk 4]
```diff
@@ -402,12 +422,15 @@ init_umath_tests(void)
     Py_DECREF(version);
 
     /* Load the ufunc operators into the module's namespace */
-    addUfuncs(d);
-
-    if (PyErr_Occurred()) {
+    if (addUfuncs(d) < 0) {
+        Py_DECREF(m);
+        if (PyErr_Occurred()) {
```
Comment (eric-wieser) [inline numpy/core/src/umath/_umath_tests.c.src]: Either:

* This test is pointless
* There is a bug in `PyUFunc_FromFuncAndDataAndSignature` which fails to set an exception correctly

I'm assuming the former, meaning you can just call `PyErr_Print()` unconditionally

File: numpy/core/src/umath/_umath_tests.c.src [hunk 5]
```diff
@@ -280,13 +280,23 @@ addUfuncs(PyObject *dictionary) {
...
                     "     \"(i),(i)->()\" \n",
                     0, inner1d_signature);
+    /*
+     * yes, this should not happen, but I (MHvK) just spent an hour looking at
+     * segfaults because I screwed up something that seemed totally unrelated.
+     */
+    if (f == NULL) {
+        return -1;
+    }
     PyDict_SetItemString(dictionary, "inner1d", f);
```
Comment (eric-wieser) [inline numpy/core/src/umath/_umath_tests.c.src]: In principle we should be checking the return value of `PyDict_SetItemString` too, but I'm happy for you to declare that out of scope for this PR.

File: numpy/core/src/umath/_umath_tests.c.src (hunk 4 above)
Comment (eric-wieser) [reply numpy/core/src/umath/_umath_tests.c.src]: I suppose my point is that `addUfuncs() < 0` is already our test to be sure that an error is set. If we returned -1 but forgot to set an error, that's a bug.

Also it seems that [`PyErr_Print` is perfectly ok with there being no error anyway](https://github.com/python/cpython/blob/42aa93b8ff2f7879282b06efc73a31ec7785e602/Python/pythonrun.c#L640-L650). So I'd say keep the PyErr_Print(), but drop the condition guarding it.

**Review by eric-wieser — APPROVED** [review]

**Comment by mattip** [comment]
Thanks Marten

**Pull request merged**